### ISSUE TITLE
Add TFCE as inference option

### ISSUE THREAD

**Issue by nicholst** [NISOx-BDI/SwE-toolbox#15]
Request from Mary Heitzeg's group at the University of Michigan

**Comment by TomMaullin** [comment]
Hi @nicholst ,

I have been looking into how this could be done and I was wondering if the following is what you had in mind;

Built into the wild bootstrap option for SwE would be a `TFCE` option. This would be calculated by;

 - First, computing the `SwE{T/F}` map parametricly.
 - Second, converting the `SwE{T/F}` map to a TFCE statistic using a function like [this](https://github.com/markallenthornton/MatlabTFCE/blob/master/matlab_tfce_transform.m) (and saving the resultant map as `swe_vox_TFCE_c{c#}.nii`). Let this original TFCE statistic map be denoted `O` and `O_v` denote it's value at voxel `v`.
 - Third, performing a bootstrap which for each bootstrap, bootstrap number $I$, does the following:
   - Calculate the `SwE{T/F}` as usual. 
   - Convert the `SwE{T/F}` into a TFCE image (denote the TFCE image from the Ith bootstrap as `B_I`).
   - For every voxel, record:
![image](https://user-images.githubusercontent.com/22241320/45231278-7076c980-b2c3-11e8-926e-cf67269445ed.png)
      where `1()` represents an indicator function.



    - Record the global maxima in the TFCE image:
![image](https://user-images.githubusercontent.com/22241320/45231372-b03db100-b2c3-11e8-8093-26eedf7bb673.png)

 - Fourth, obtaining:
   - The uncorrected TFCE P values from:
![image](https://user-images.githubusercontent.com/22241320/45231567-4245b980-b2c4-11e8-9c18-4cba5c7d600a.png)
   - The FWE voxelwise corrected TFCE P values from looking at the distribution of the recorded maxima.
   - The FDR voxelwise corrected TFCE P values from performing: `spm_P_FDR` on the uncorrected TFCE P values.

The only additional choices the user should have is the variable `dh` for computing the TFCE integral.

The resultant output of this process should then be the maps:
 - `swe_vox_TFCE_c{c#}.nii` - The parametric TFCE image.
 - `swe_vox_TFCE_lp-WB_c{c#}.nii` - The TFCE uncorrected p value image.
 - `swe_vox_TFCE_lpFWE-WB_c{c#}.nii` - The TFCE voxelwise FWE corrected p value image.
 - `swe_vox_TFCE_lpFDR-WB_c{c#}.nii` - The TFCE voxelwise FDR corrected p value image

As well as the usual parametric maps.

The display should then only have columns for:
 - set-level: P and c values.
 - cluster-level: Nothing.
 - peak-level: TFCE FWE corrected p values, TFCE FDR corrected p values, the TFCE statistic, TFCE uncorrected p values
 - mm coordinates of peaks

Is this close to what you had in mind? Or should this be done separate from the bootstrap using a different permutation method?

**Comment by nicholst** [comment]
Hi @TomMaullin ,

Right!  I had forgotten about MatlabTFCE and *amazed* at the concision of that function https://github.com/markallenthornton/MatlabTFCE/blob/master/matlab_tfce_transform.m ... we only need that one function!

You've really got your head wrapped around it! Just a few things to note:

1. The TFCE transform is constructed for z-score like-quantities.  That means that F-test values will need to be transformed to Z's via a probability integral transform... look it up; it basically a way to transform any r.v. via p-values, like 
```matlab
d1=3;d2=100;
SomeFs=frnd(d1,d2,1000,1);
SomeSz=-norminv(fcdf(SomeFs,d1,d2,'upper'));
```
where we've carefully used the complementary (upper tail) CDF to avoid extreme F-values creating near CDF values that underflow to 1.0.

2. "TFCE" is the statistic name/type, so the naming will be: `swe_TFCE_...` otherwise your naming is perfect.

Otherwise, you've nailed it!  Once we do it here, it would be great to add it to SnPM, as per here: https://github.com/SnPM-toolbox/SnPM-devel/issues/54

**Comment by TomMaullin** [comment]
Ah great! Ah right of course yes! The only two things I am still unsure on are as follows:

 - How would you want this to be implemented into the batch window? Should it be listed under WB if it is based on a wild bootstrap? If not then the user must not be able to select both WB and TFCE at the same time so maybe the batch window should have an "Alternative inference" option where WB and TFCE are both options the user can select?

 - Just to double check; in the TFCE paper a permutation test is used to derive the uncorrected and FwE p-values but in the above suggestion we would be using wild bootstrap? As these two methods use different assumptions I thought I had best double check if deriving TFCE p values using a wild boostrap is a valid thing to do? Should I use a permutation test instead?

**Comment by nicholst** [comment]
Did our meeting sort all these out?  As TFCE will be considerably slower than cluster inference, I think it does need to be a 'switch' somewhere. "TFCE inference: Off" vs "TFCE inference: On" with help explaining
> Enabling TFCE inference will replace cluster inference with Threshold-Free Cluster Enhancement.  A cluster-forming threshold can no longer be specified in the results, and only a FWE-corrected P-value can be used to set the threshold.

Is this good?  *Or* maybe we want to have cluster inference recorded, and *if* TFCE was computed, the user can choose at results time to show TFCE or cluster inference???

**Comment by nicholst** [comment]
Yes, this looks good.

As for options, we could include the H and E values... some people want to
mess with those.

**Comment by TomMaullin** [comment]
Ah right, yes! Should `dh` also be an input?

**Comment by TomMaullin** [comment]
**Questions/Updates**

Hi @nicholst ,

This is just a quick further questions/updates post. Currently, the code has been updated to compute TFCE uncorrected and FWE-corrected P-values for T statistics (and almost for F statistics also) for volumetric input. Below is the stages I plan on implementing the TFCE in (I will tick stages off as I complete them):

 - [ ] Implement TFCE into `swe_cp_WB`:
    - [x] Create TFCE uncorrected and fwe p value images for volumetric T tests.
    - [ ] Create TFCE uncorrected and fwe p value images for volumetric F tests.
    - [ ] Create TFCE uncorrected and fwe p value images for surface T tests.
    - [ ] Create TFCE uncorrected and fwe p value images for surface F tests.
 - [ ] Update the GUI with WB options.
 - [ ] Update the results display to display WB if WB has been selected.
 - [ ] Update the Travis tests with updated SwE structures and new use cases.

I did have the following questions regarding the TFCE update, however:

 - Just to double check, should I also be implementing TFCE for surface-based tests? I believe I can do this, I just wanted to check it would be valid/correct.
 - For a T test, should the TFCE statistic be derived from the T statistic or the equivalent Z statistic? I have currently implemented it for the T statistic but it would be relatively quick to change. I only ask as the F test will have to be based on an equivalent Z statistic so for consistency it seems it would make sense to use the equivalent Z for a T test.
 - When using an equivalent Z for the F contrast, as you suggest in the previous comments, what should I set the degrees of freedom to? Currently I am just using `[1, 1]`.

**Comment by nicholst** [comment]
Quick answers:

Surface would be great, but I didn't think we had that functionality at all
(e.g. for clusters); I'd put it at the bottom of the priority list (do it
last).
Yes, I think making an equivalent Z the basis of all TFCE makes sense.
You need to get the F DF from the test... or am I missing something?

**Comment by TomMaullin** [comment]
Ah okay, sorry yes I should have been clearer - I thought I had best double check - should I use the spatially varying df or the nominal df?

**Comment by TomMaullin** [comment]
This issue has now been addressed by PR #48
